### PR TITLE
adding knative9 vendor changes + duck API changes

### PR DESCRIPTION
**Pull request by animeshsingh**
adding knative9 vendor dependencies + upgrading duck API from beta. This upgrades KFServing to Knative 9 + Istio 1.3.1

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/kfserving/543)

### CODE REVIEW

**Comment by animeshsingh** [comment]
@jinchihe belive we cannot have this tested through e-2-e given its on kn8. So best resort is manual testing?

**Comment by jinchihe** [comment]
@animeshsingh so do you think if need to upgrade knative verison to 0.9.0 in e2e testing beflow?

https://github.com/kubeflow/kfserving/blob/cc8f507374b690c0c83cc7311ed698fcb69bd1a4/test/scripts/run-e2e-tests.sh#L29

**Comment by animeshsingh** [comment]
> I think we should update related code and UT case accordingly.
> And do you think if need to upgrade knative verison to 0.9.0 in e2e testing beflow?
> 
> https://github.com/kubeflow/kfserving/blob/cc8f507374b690c0c83cc7311ed698fcb69bd1a4/test/scripts/run-e2e-tests.sh#L29

The ideal case would be Istio 1.3.1 and Knative 0.9 for e-2-e. What are Unit tests which can be run individually?

**Comment by Tomcli** [comment]
Looks like knative 0.9 vendor deprecated few apis that were used in kfserving. (e.g. revisionspec in knative 0.9 is migrated from v1beta1 to v1, https://github.com/kubeflow/kfserving/pull/543/files#diff-e3b0b43978a9022037342ea8fd482e05) 

Should we include those api migration in this PR as well? or should we push those in a separate PR?

**Comment by animeshsingh** [comment]
> Looks like knative 0.9 vendor deprecated few apis that were used in kfserving. (e.g. revisionspec in knative 0.9 is migrated from v1beta1 to v1, https://github.com/kubeflow/kfserving/pull/543/files#diff-e3b0b43978a9022037342ea8fd482e05)
> 
> Should we include those api migration in this PR as well? or should we push those in a separate PR?
Vendor changes typically go in one PR. Please do another one as we test through it, and all of both of them can go in at once

**Comment by animeshsingh** [comment]
@jinchihe coming back to the unit tests, are there tests which @Tomcli can run to verify?

**Comment by jinchihe** [comment]
@Tomcli @animeshsingh if we do not include those api migration in this PR, we may cannot pass the CI test, if we create a separate PR, I'm not sure if the "separated PR" can pass the CI test without the knative 0.9.  If the unit test in CI cannot pass, the PR cannot be merged, that's unimaginable if merge the big changes manually...
@animeshsingh I'm not sure the case list, but why not execute all unit test cases, that should be finished less than 2-3 minutes.

**Comment by animeshsingh** [comment]
@jinchihe that's a guiding principle we are following, where vendor changes live in a separate PR. Please look at the guidelines. And in the past in cases like, a manual merge is done.

And again, as I have asked, where are the unit test cases listed? I have no qualms running them - please point it to me.

**Comment by Tomcli** [comment]
Hi @jinchihe, I realized the kfserving controller costs a lot of memory to initialize and often time I see lots of OOMKilled. Then I try to remove the memory limit and see some initial memory spike. Is there any performance test that I can run to check any possible memory leak? Thank you in advance. 
```
(py3) Tommys-MacBook-Pro:postprocessing tommyli$ kubectl get pod -n kfserving-system
NAME                             READY   STATUS    RESTARTS   AGE
kfserving-controller-manager-0   2/2     Running   0          19s
(py3) Tommys-MacBook-Pro:postprocessing tommyli$ kubectl top pod -n kfserving-system
NAME                             CPU(cores)   MEMORY(bytes)
kfserving-controller-manager-0   7m           923Mi
```

**Comment by jinchihe** [comment]
@Tomcli I think no performance test so far but we may add that in next release, I would suggest you set the resources section as below to work around this.

```
        resources:
          requests:
            cpu: 0.1
            memory: 200Mi                       
          limits:
            memory: 200Mi
```

**Comment by jinchihe** [comment]
@animeshsingh sorry, what exactly do you want to know for "unit test cases"? that's in our go code, and we can run unit test individually following below steps:
```
curl -L -O https://storage.googleapis.com/kubebuilder-release/kubebuilder_master_linux_amd64.tar.gz
tar -zxvf kubebuilder_master_linux_amd64.tar.gz
mv kubebuilder_master_linux_amd64 kubebuilder && mv kubebuilder /usr/local/
export PATH=$PATH:/usr/local/kubebuilder/bin:${GOPATH}/bin
mkdir -p ${GOPATH}/src/github.com/kubeflow/
Copy you kfserving code to ${GOPATH}/src/github.com/kubeflow/
cd ${GOPATH}/src/github.com/kubeflow/kfserving
make test
```

**Comment by hougangliu** [comment]
@animeshsingh do you mean https://github.com/kubeflow/kfserving/blob/master/test/e2e/test_tensorflow_inference_service.py about CI tests？

**Comment by animeshsingh** [comment]
@hougangliu perfect. Yes - test 2 is what we need to run. Installation/configuration etc is anyways out of the scope of this
https://github.com/kubeflow/kfserving/blob/master/test/e2e/test_tensorflow_inference_service.py
cc @Tomcli

**Comment by Tomcli** [comment]
@animeshsingh I'm able to run the `test_tensorflow_inference_service.py` on our IBM Cloud with #550 using knative 0.9, Istio 1.3.1, Kubernetes 1.14.6. Is that all the tests we have?

**Comment by animeshsingh** [comment]
@Tomcli for the 2nd part of e-2-e, yes that's all we have. Once this is merged, we may discover edge cases which we resolve as we hit.

**Comment by animeshsingh** [comment]
Awesome - all the tests passed.  Thanks @jinchihe for upgrading the tests to knative 0.9 + Istio 1.3.1, and thanks @yuzisun for making them comprehensive

Screen Shot 2019-11-22 at 6 24 09 PM

**Comment by yuzisun** [comment]
kfserving unit tests needs to upgrade too, it is still using the alpha1 crd.
https://github.com/kubeflow/kfserving/blob/master/test/crds/knative_service.yaml#L24

**Comment by jinchihe** [comment]
Better to update the knative serving version unit tests that piont out by @yuzisun above in this PR? Thanks @animeshsingh

**Comment by Tomcli** [comment]
@jinchihe If we are going to change knative serving version unit tests to v1, then we probably also want to update the knative route and configuration to v1. When we are making those changes, it requires us to migrate the Kubernetes api version from 1.12.3 to 1.14.x as stated in the knative v0.9 release. (https://github.com/knative/serving/releases/tag/v0.9.0) This means that there will be a significant amount of vendor changes and we might lose the backward capability with knative 0.8 and below because they don't support the latest v1 api. 

I suggest we should merge this PR with the v1alpha1 CRD first. Then we can bring these discussions on a separate PR for the knative v1 crd migration. Does this sound good to you? 

Thanks.

**Comment by animeshsingh** [comment]
+1. Let's open an issue for API changes, and tackle that in a separate PR

**Comment by jinchihe** [comment]
> I suggest we should merge this PR with the v1alpha1 CRD first. Then we can bring these discussions on a separate PR for the knative v1 crd migration. Does this sound good to you?
> 
@Tomcli Sounds good! Thanks.

File: Gopkg.toml [hunk 1]
```diff
@@ -46,7 +46,7 @@ version="v1.4.7"
 
 [[constraint]]
   name = "knative.dev/serving"
-  version = "=v0.8.0"
+  version = "=v0.9"
```
Comment (ellis-bigelow) [inline Gopkg.toml]: Hate to say this, but any reason we cant jump to 0.10?

Comment (animeshsingh) [reply Gopkg.toml]: This is one is tested with test infrastructure et all pointed to Istio 1.3.1 and KNative 0.9. I will move this in, catch any edge cases in the meanwhile and follow-on quickly with 0.10

Comment (yuzisun) [reply Gopkg.toml]: I am about to send a PR to upgrade kubeflow knative manifests to 0.10, is it a quick change that just update to v0.10 here and e2e test to install 0.10?

**Comment by yuzisun** [comment]
> @jinchihe If we are going to change knative serving version unit tests to v1, then we probably also want to update the knative route and configuration to v1. When we are making those changes, it requires us to migrate the Kubernetes api version from 1.12.3 to 1.14.x as stated in the knative v0.9 release. (https://github.com/knative/serving/releases/tag/v0.9.0) This means that there will be a significant amount of vendor changes and we might lose the backward capability with knative 0.8 and below because they don't support the latest v1 api.
> 
> I suggest we should merge this PR with the v1alpha1 CRD first. Then we can bring these discussions on a separate PR for the knative v1 crd migration. Does this sound good to you?
> 
> Thanks.

We no longer directly create knative route and configuration, also although we are all changing to use v1 service, we are still registering the v1alpha crd in controller.  I think both here(https://github.com/kubeflow/kfserving/blob/master/cmd/manager/main.go#L69) and unit test need to be changed to use v1.

File: Gopkg.toml (hunk 1 above)
Comment (animeshsingh) [reply Gopkg.toml]: Here:
1. IBM KNative service is still on 0.8, and they are releasing 0.9 in Dec sometime. 
2. GKE is also running on 0.9 currently
2. Our internal environments are either 0.9, or 0.8, and at this point, I would need to verify the feasibility of 1.0 before taking a final call
3. Also took 7-8 days to coordinate, get test infra and environments updated, test both manually and in automated fashion. 

I will roll this in, and in next one I will jump to 0.10 immediately in form of a PR, have @jinchihe upgrade the test infra in another PR, and we also update the environments on our side to run tests and also verify manually. 

In parallel, I would get the feasibility date from IBM and GCP side, before taking a final call to jump to 0.10 immediately
